### PR TITLE
Move offerings endpoints to a JSON RPC server.

### PR DESCRIPTION
**Pull request by dzeckelev**
Resolves BV-759.

### CODE REVIEW

File: ui/errors.go [hunk 1]
```diff
@@ -38,6 +43,11 @@ var errMsgs = errors.Messages{
 	ErrProductNotFound:         "product not found",
 	ErrInvalidTemplateType:     "invalid template type",
 	ErrMalformedTemplate:       "malformed template",
+	ErrSmallDeposit:            "deposit is too small",
```
Comment (ababo) [inline ui/errors.go]: ErrDepositTooSmall

File: ui/errors.go [hunk 2]
```diff
@@ -38,6 +43,11 @@ var errMsgs = errors.Messages{
...
+	ErrBadUnitPriceRange:       "bad unit price range",
+	ErrBadUnitType:             "bad unit type",
+	ErrBillingType:             "bad billing type",
+	ErrBadOfferingStatusAction: "bad action for change a offering status ",
```
Comment (ababo) [inline ui/errors.go]: "bad offering status action"

File: ui/misc.go [hunk 3]
```diff
@@ -16,7 +16,12 @@ const (
 			AND status = 'msg_channel_published'
 			AND NOT is_local
 			AND current_supply > 0
-			AND agent NOT IN (SELECT eth_addr FROM accounts)`
+			AND agent NOT IN (SELECT eth_addr FROM accounts)
```
Comment (ababo) [inline ui/misc.go]: You use this constant only once. Are you sure it should be defined separately?

File: ui/offering.go [hunk 4]
```diff
@@ -1,23 +1,51 @@
...
 // AcceptOfferingResult is an AcceptOffering result.
 type AcceptOfferingResult struct {
 	Channel string `json:"channel"`
 }
 
-// AcceptOffering initiates JobClientPreChannelCreate job and subscribes to job
-// results of the corresponding flow.
+// GetOfferingsResult is a GetClientOfferings and a GetAgentOfferings result.
+type GetOfferingsResult struct {
+	Offerings []data.Offering `json:"offerings"`
```
Comment (ababo) [inline ui/offering.go]: Please remove all the Result-structures.

File: ui/offering.go [hunk 5]
```diff
@@ -29,18 +57,306 @@ func (h *Handler) AcceptOffering(password, account, offering string,
...
+
+	return nil
+}
+
+func (h *Handler) getClientOfferingsConditions(
+	agent string, minUnitPrice, maxUnitPrice uint64,
+	country []string) (conditions string, arguments []interface{}) {
+
+	count := 1
+	operator := "AND"
```
Comment (ababo) [inline ui/offering.go]: Why can't we inline the operator? You don't alter it throughout the code.

File: ui/offering_test.go [hunk 6]
```diff
@@ -47,3 +53,340 @@ func TestAcceptOffering(t *testing.T) {
...
+	if res == nil {
+		if expNumber > 0 {
+			t.Fatalf("number of offerings got 0,"+
+				" expect %d", expNumber)
+		}
+		return
+	}
+
+	if len(res.Offerings) != expNumber {
+		t.Fatalf("number of offerings got %d, expect %d",
```
Comment (ababo) [inline ui/offering_test.go]: expected: , got:

File: ui/offering_test.go [hunk 7]
```diff
@@ -47,3 +53,340 @@ func TestAcceptOffering(t *testing.T) {
...
+
+	o = newTestOffering(fxt)
+	o.UnitName = ""
+	offerings = append(offerings, o)
+
+	o = newTestOffering(fxt)
+	o.UnitType = ""
+	offerings = append(offerings, o)
+
+	return offerings
```
Comment (ababo) [inline ui/offering_test.go]: Use some shortening helper-function, lot's of similar blocks. Please consider to shorten other test cases as well.

**Comment by dzeckelev** [comment]
@ababo updated.

File: doc/ui/rpc.md [hunk 8]
```diff
@@ -150,62 +149,93 @@ This document describes a UI JSON RPC API located in "ui" namespace.
...
+*Description*: Get offerings available for a agent.
 
 *Parameters*:
 1. Password (string)
-2. Object with keys as setting names and values as setting values (object)
+2. Product id (string)
+3. Offering status (string, can be `empty`, `register` or `remove`)
 
-*Result*: None.
+*Result (array of `data.Offering` objects)*: offering available for a agent.
```
Comment (ababo) [inline doc/ui/rpc.md]: offerings.

File: doc/ui/rpc.md [hunk 9]
```diff
@@ -150,62 +149,93 @@ This document describes a UI JSON RPC API located in "ui" namespace.
...
 
+#### Get Offerings For Client
 
-### Templates
+*Method*:	`getClientOfferings`
 
-#### Get Templates
+*Description*: Get active offerings available for a client.
```
Comment (ababo) [inline doc/ui/rpc.md]: Get active client offerings.

File: doc/ui/rpc.md [hunk 10]
```diff
@@ -150,62 +149,93 @@ This document describes a UI JSON RPC API located in "ui" namespace.
...
+*Result (string)*: id of offering to be created.
 
-#### Update Settings
+#### Get Offerings For Agent
 
-*Method*:	`updateSettings`
+*Method*:	`getAgentOfferings`
 
-*Description*: Update existing settings.
+*Description*: Get offerings available for a agent.
```
Comment (ababo) [inline doc/ui/rpc.md]: Get active agent offerings.

File: doc/ui/rpc.md [hunk 11]
```diff
@@ -150,62 +149,93 @@ This document describes a UI JSON RPC API located in "ui" namespace.
...
-*Method*:	`getTemplates`
+*Parameters*:
+1. Password (string)
+2. Agent ethereum address (string)
+3. Minimum unit price (number)
+4. Maximum unit price (number)
+5. Country codes ISO 3166-1 alpha-2 (array of string)
 
-*Description*: Get templates.
+*Result (array of `data.Offering` objects)*: offering available for a client.
```
Comment (ababo) [inline doc/ui/rpc.md]: : offerings.

File: doc/ui/rpc.md [hunk 12]
```diff
@@ -245,38 +275,76 @@ This document describes a UI JSON RPC API located in "ui" namespace.
...
+1. Password (string)
+2. Template type (string, can be `offer` or `access`)
+
+*Result (object)*:
+- `templates` (array of `data.Template` objects) - returned templates.
+
+
+### Transactions
+
+#### Get Transactions
```
Comment (ababo) [inline doc/ui/rpc.md]: Get Ethereum Trasnactions

File: doc/ui/rpc.md [hunk 13]
```diff
@@ -245,38 +275,76 @@ This document describes a UI JSON RPC API located in "ui" namespace.
...
+
+*Method*:	`getEthTransactions`
+
+*Description*: Get Ethereum transactiosn.
```
Comment (ababo) [inline doc/ui/rpc.md]: transactions

File: doc/ui/rpc.md [hunk 14]
```diff
@@ -245,38 +275,76 @@ This document describes a UI JSON RPC API located in "ui" namespace.
...
+*Method*:	`getTemplates`
+
+*Description*: Get templates.
+
+*Parameters*:
+1. Password (string)
+2. Template type (string, can be `offer` or `access`)
+
+*Result (object)*:
+- `templates` (array of `data.Template` objects) - returned templates.
```
Comment (ababo) [inline doc/ui/rpc.md]: \- templates.
Please remove the returned object as we do for the rest of methods.

File: ui/offering.go [hunk 15]
```diff
@@ -24,18 +42,310 @@ func (h *Handler) AcceptOffering(password, account, offering string,
...
 
 	return &rid, nil
 }
+
+// ChangeOfferingStatus initiates JobAgentPreOfferingMsgBCPublish,
+// JobAgentPreOfferingPopUp or JobAgentPreOfferingDelete job,
+// depending on a selected action.
+func (h *Handler) ChangeOfferingStatus(
+	password, offering, action string, gasPrice uint64) error {
+	logger := h.logger.Add("method", "ChangeOfferingStatus",
```
Comment (ababo) [inline ui/offering.go]: Action is missing

File: ui/offering.go [hunk 16]
```diff
@@ -24,18 +42,310 @@ func (h *Handler) AcceptOffering(password, account, offering string,
...
+	}
+
+	conditions = fmt.Sprintf(format, conditions, activeOfferingCondition)
+	return conditions, arguments
+}
+
+// GetClientOfferings returns active offerings available for a client.
+func (h *Handler) GetClientOfferings(
+	password, agent string, minUnitPrice, maxUnitPrice uint64,
+	country []string) ([]data.Offering, error) {
```
Comment (ababo) [inline ui/offering.go]: countries

File: ui/offering.go [hunk 17]
```diff
@@ -24,18 +42,310 @@ func (h *Handler) AcceptOffering(password, account, offering string,
...
+	logger := h.logger.Add("method", "GetClientOfferings",
+		"agent", agent, "minUnitPrice", minUnitPrice,
+		"maxUnitPrice", maxUnitPrice, "country", country)
```
Comment (ababo) [inline ui/offering.go]: countries

File: ui/offering_test.go [hunk 18]
```diff
@@ -47,3 +66,576 @@ func TestAcceptOffering(t *testing.T) {
...
+	fxt *fixture, assertMatchErr func(error, error), agent string) {
+	_, err := handler.GetClientOfferings(
+		"wrong-password", "", 0, 0, nil)
+	assertMatchErr(ui.ErrAccessDenied, err)
+
+	lowPrice := fxt.Offering.UnitPrice - 10
+	price := fxt.Offering.UnitPrice
+	highPrice := fxt.Offering.UnitPrice + 10
+
+	for _, v := range []struct {
```
Comment (ababo) [inline ui/offering_test.go]: Please define a structure and move the initialisers into a dedicated var and flatten them. The same for the rest.

**Comment by dzeckelev** [comment]
@ababo updated.

File: doc/ui/rpc.md [hunk 19]
```diff
@@ -150,62 +149,93 @@ This document describes a UI JSON RPC API located in "ui" namespace.
...
-#### Get Templates
+*Description*: Get active client offerings.
 
-*Method*:	`getTemplates`
+*Parameters*:
+1. Password (string)
+2. Agent ethereum address (string)
+3. Minimum unit price (number)
+4. Maximum unit price (number)
+5. Country codes ISO 3166-1 alpha-2 (array of string)
```
Comment (ababo) [inline doc/ui/rpc.md]: array of strings

File: ui/offering.go [hunk 20]
```diff
@@ -24,18 +42,310 @@ func (h *Handler) AcceptOffering(password, account, offering string,
...
+			conditions = condition
+		} else {
+			conditions = fmt.Sprintf(
+				"%s AND %s", conditions, condition)
+		}
+		args = append(args, status)
+		index++
+	}
+
+	prefix := "WHERE "
```
Comment (ababo) [inline ui/offering.go]: Are you sure we need that? Why not to inline?

Comment (dzeckelev) [reply ui/offering.go]: I use this variable in several places. This variable slightly reduces duplication of code.

**Pull request merged**